src/core/mainloop/connection.c

```diff
@@ -2759,20 +2844,54 @@ retry_all_listeners(smartlist_t *replaced_conns,
   if (retry_listener_ports(listeners,
                            get_configured_ports(),
                            new_conns,
+                           replacements,
                            close_all_noncontrol) < 0)
     retval = -1;
 
+#ifdef ENABLE_LISTENER_REBIND
+  SMARTLIST_FOREACH_BEGIN(replacements, listener_replacement_t *, r) {
+    int addr_in_use = 0;
+    int skip = 0;
+
+    tor_assert(r->new_port);
+    tor_assert(r->old_conn);
+
+    connection_t *new_conn =
+     connection_listener_new_for_port(r->new_port, &skip, &addr_in_use);
+    connection_t *old_conn = r->old_conn;
+
+    if (skip)
+      continue;
+
+    connection_close_immediate(old_conn);
+    connection_mark_for_close(old_conn);
```

Comment:
Are both needed here?

Reply:
I believe so - we want to close the old connection right away, without trying to send any data from its write buffer. It is generally done like this in connection.c when closing no longer configured or failed listeners.

See:
* https://gitweb.torproject.org/tor.git/tree/src/core/mainloop/connection.c#n2773
* https://gitweb.torproject.org/tor.git/tree/src/core/mainloop/connection.c#n3426